### PR TITLE
Add distillation loop

### PR DESCRIPTION
**Pull request by danieldk**
## Description

This change adds the distillation loop. The loop is based on the training loop, but adjusted to do distillation. The change also introduces a new `init_nlp_student` function used to initialize the student pipe. Both are tested through a test that first trains a teacher model using the `train` loop and then distills a student using the `distill` loop.

### Types of change

Feature.

## Checklist

- [x] I confirm that I have the right to submit this contribution under the project's MIT license.
- [x] I ran the tests, and all new and existing tests passed.
- [x] My changes don't require a change to the documentation, or if they do, I've added all required information.

### CODE REVIEW

File: spacy/training/loop.py [hunk 1]
```diff
@@ -139,6 +266,126 @@ def save_checkpoint(is_best):
...
+            before_update_args = {"step": step, "epoch": epoch}
+            before_update(student, before_update_args)
+        dropout = dropouts(optimizer.step)
+        for subbatch in subdivide_batch(batch, accumulate_gradient):
+            student.distill(
+                teacher,
+                subbatch,
+                drop=dropout,
+                losses=losses,
+                sgd=None,
```
Comment (shadeMe) [inline spacy/training/loop.py:343]: This should be `False`, right? So, that we don't end up calling `finish_update` twice.

File: spacy/training/initialize.py [hunk 2]
```diff
@@ -101,6 +94,102 @@ def init_nlp(config: Config, *, use_gpu: int = -1) -> "Language":
...
+        raise ConfigValidationError(
+            desc=Errors.E897.format(field="distillation.corpus", type=type(D["corpus"]))
+        )
+    if not isinstance(T["dev_corpus"], str):
+        raise ConfigValidationError(
+            desc=Errors.E897.format(
+                field="training.dev_corpus", type=type(T["dev_corpus"])
+            )
+        )
+    distill_corpus, dev_corpus = resolve_dot_names(config, dot_names)
```
[the remark this answers is not in the document]
Comment (danieldk) [reply spacy/training/initialize.py]: Fixed.

File: spacy/training/loop.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (danieldk) [reply spacy/training/loop.py]: Yeah, though `Language.distill` had the flawed implementation that did not accept `False`, so updated that method as well to support the proper semantics.

File: spacy/training/loop.py [hunk 3]
```diff
@@ -139,6 +266,126 @@ def save_checkpoint(is_best):
...
+    with each iteration yielding a tuple `(batch, info, is_best_checkpoint)`,
+    where info is a dict, and is_best_checkpoint is in [True, False, None] --
+    None indicating that the iteration was not evaluated as a checkpoint.
+    The evaluation is conducted by calling the evaluate callback.
+
+    Positional arguments:
+        teacher (Language): The teacher pipeline to distill from.
+        student (Language): The student pipeline to distill into.
+        optimizer: The optimizer callable.
+        distill_data (Iterable[Batch]): A generator of batches, with the distillation
```
[the remark this answers is not in the document]
Comment (danieldk) [reply spacy/training/loop.py]: Yeah, that seems like a leftover from spaCy from before `Example`? I removed the sentence altogether. I also made similar changes to `train_while_improving` while we are at it.

File: spacy/training/loop.py [hunk 4]
```diff
@@ -139,6 +266,126 @@ def save_checkpoint(is_best):
...
+    dropout: float,
+    eval_frequency: int,
+    accumulate_gradient: int,
+    max_steps: int,
+    exclude: List[str],
+    annotating_components: List[str],
+    before_update: Optional[Callable[["Language", Dict[str, Any]], None]],
+    student_to_teacher: Dict[str, str],
+):
+    """Train until an evaluation stops improving. Works as a generator,
```
[the remark this answers is not in the document]
Comment (danieldk) [reply spacy/training/loop.py]: Updated.

File: spacy/training/loop.py [hunk 5]
```diff
@@ -264,7 +511,10 @@ def train_while_improving(
 
 def subdivide_batch(batch, accumulate_gradient):
     batch = list(batch)
-    batch.sort(key=lambda eg: len(eg.predicted))
+    if isinstance(batch, Example):
```
[the remark this answers is not in the document]
Comment (danieldk) [reply spacy/training/loop.py]: Fixed.

File: spacy/training/loop.py [hunk 6]
```diff
@@ -21,6 +22,132 @@
...
+        printing, set to io.StringIO.
+
+    RETURNS (tuple): The final student nlp object and the path to the exported
+        student model.
+    """
+    # We use no_print here so we can respect the stdout/stderr options.
+    msg = Printer(no_print=True)
+    # Create iterator, which yields out info after each optimization step.
+    config = student.config.interpolate()
+    if config["training"]["seed"] is not None:
```
[the remark this answers is not in the document]
Comment (svlandeg) [reply spacy/training/loop.py:56]: Alternatively, we could also move those helper functions to `util.py`?

[the remark this answers is not in the document]
Comment (danieldk) [reply spacy/training/loop.py]: The helper functions are now used in `loop.py` as well and I moved them to `spacy.util`.

**Review by shadeMe — APPROVED** [review]

File: spacy/tests/training/test_loop.py [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+        {"sent_starts": SENT_STARTS},
+    ),
+    (
+        "She likes purple eggs. They hate ham. You like yellow eggs.",
+        {"sent_starts": SENT_STARTS},
+    ),
+]
+
+
+@pytest.mark.slow
```
Comment (svlandeg) [inline spacy/tests/training/test_loop.py:87]: As there's a new slow test, best to have our bot run the slow test suite on the PR.

**Comment by shadeMe** [comment]
@explosion-bot please test_slow

**Pull request merged**